api/envoy/type/tracing/v2/custom_tag.proto

```diff
@@ -0,0 +1,105 @@
+syntax = "proto3";
+
+package envoy.type.tracing.v2;
+
+option java_outer_classname = "CustomTagProto";
+option java_multiple_files = true;
+option java_package = "io.envoyproxy.envoy.type.tracing.v2";
+
+import "envoy/type/metadata/v2/metadata.proto";
+
+import "validate/validate.proto";
+
+// [#protodoc-title: Custom Tag]
+
+// Describes custom tags for the active span.
+// [#next-free-field: 6]
+message CustomTag {
+  // Literal type custom tag with static value for the tag value.
+  message Literal {
+    // Static literal value to populate the tag value.
+    string value = 1 [(validate.rules).string = {min_bytes: 1}];
+  }
+
+  // Environment type custom tag with environment name and default value.
+  message Environment {
+    // Environment variable name to obtain the value to populate the tag value.
+    string name = 1 [(validate.rules).string = {min_bytes: 1}];
+
+    // When the environment variable is not found,
+    // the tag value will be populated with this default value if specified,
+    // otherwise no tag will be populated.
+    string default_value = 2;
+  }
+
+  // Header type custom tag with header name and default value.
+  message Header {
+    // Header name to obtain the value to populate the tag value.
+    string name = 1 [(validate.rules).string = {min_bytes: 1}];
+
+    // When the header does not exist,
+    // the tag value will be populated with this default value if specified,
+    // otherwise no tag will be populated.
+    string default_value = 2;
+  }
+
+  // Metadata type custom tag using
+  // :ref:`MetadataKey <envoy_api_msg_type.metadata.v2.MetadataKey>` to retrieve tag value
+  // from :ref:`Metadata <envoy_api_msg_core.Metadata>`.
+  // The value will be formatted in the following way depending on the type:
+  //
+  // * `boolean`: populate with ``true`` or ``false`` string;
+  // * `number`: populate with number string;
+  // * `string`: populate with the string itself, an empty tag value
+  //   will be populated if it is empty;
+  // * `list`/`struct`: populate with the corresponding JSON string
+  //   converted from the protobuf message;
+  // * otherwise means that no valid metadata is found.
+  message Metadata {
+    enum Source {
```

Comment:
For future proofing, please move this to a oneof with a required field, using empty messages for each type. This will give us flexibility later if needed. Also, nit: Start all sentences with capital letters, etc.